design/baremetal-operator/bmh-v1beta1.md

```diff
@@ -0,0 +1,126 @@
+<!--
+ This work is licensed under a Creative Commons Attribution 3.0
+ Unported License.
+
+ http://creativecommons.org/licenses/by/3.0/legalcode
+-->
+
+# BareMetalHost API v1beta1
+
+## Status
+
+One of: implementable
+
+## Summary
+
+This design proposes making certain backward incompatible changes to the
+BareMetalHost API and publishing the result as version `v1beta1`.
+
+## Motivation
+
+The BareMetalHost API is currently stuck at `v1alpha1`. Meanwhile, several
+incompatible changes are needed based on the recent development both in Metal3
+itself and in Ironic.
+
+For example, Ironic is planning a complete removal of MD5 support for
+checksums.
+
+### Goals
+
+- BareMetalHost API version `metal3.io/v1beta1`.
+
+### Non-Goals
+
+- Stabilize the BareMetalHost API.
+
+## Proposal
+
+Copy the current API types definition into a new package `v1beta1` with some
+modifications.
+
+## Design Details
+
+The following addition should be done *before* the version is created:
```

Comment:
I would suggest actually creating v1alpha2 as a direct copy of the current v1alpha1. Then implement these changes in v1alpha2. When done, make a release. If we are really happy/confident with it, rename v1alpha2 -> v1beta1 before the release. But I think it would be good to actually do a v1alpha2 release first. It can anyway be promoted to v1beta1 without changes if it proves good.

Reply:
Why? You're essentially asking for double the work. We're definitely not in the alpha stage (for a long time already).

Reply:
To quote https://kubernetes.io/docs/reference/using-api/#api-versioning about alpha versions:

> The software may contain bugs. Enabling a feature may expose bugs.
> Support for an alpha API may be dropped at any time without notice.
> The API may change in incompatible ways in a later software release without notice.
> The software is recommended for use only in short-lived testing clusters, due to increased risk of bugs and lack of long-term support.

This is obviously false for us. Compare with beta definition:

> The software is well tested. Enabling a feature is considered safe.
> The support for a feature will not be dropped, though the details may change.
> The schema and/or semantics of objects may change in incompatible ways in a subsequent beta or stable API version. When this happens, migration instructions are provided. Adapting to a subsequent beta or stable API version may require editing or re-creating API objects, and may not be straightforward. The migration may require downtime for applications that rely on the feature.

Reply:
> You're essentially asking for double the work.

I think 90% of the work is in getting things set up so that we _can_ add a new version. After that, subsequent new versions should be comparatively easy if we've done it right.

Reply:
> If we immediately create v1beta1, work on it and then make a release, then any kind of mistakes in the API would require v1beta2 to fix. 

Same applies to alpha, no differences here.

Our software is not alpha. I suspect you're using some definition of "alpha" and "beta" that do not match the kubernetes one that I quoted above. Implicit in what you suggest seems to be an assumption that beta is stable and cannot/should not be changed.

This is not what the kubernetes docs say.

Alpha = do not use in production. Is it what you expect of the new API version (let alone the current one)? At least in OpenShift we've been using BareMetalHost API in production for years already.

Beta = can be used in production but is subject to further changes. This is the state of things the way I see it *already*. I'd probably have more understanding if you suggested that we rename the current v1alpha1 to v1beta1 first, then introduce v1beta1. But that's additional work for both the consumers and myself, and I don't have infinite cycles to juggle with versions just for the sake of it.

Reply:
I get what you are saying about alpha/beta, but the fact is that the current API is labeled as alpha. This means we have the chance to work on it still under that label. I'm also fine releasing it as v1beta1 in the current state for the reasons you mentioned. That makes sense. Some people seem to think it is a failure if the beta cannot be promoted to v1 without changes. I'm not one of them, but I want to make sure that it is known before we willingly take a path that would require a v1beta2.

Since the plan is to make changes, I think it makes much more sense to be careful about them. I.e. call it alpha since the current version is anyway called alpha. This gives us the chance to iron out any issues before we release it on the world as production ready. It would be very embarrassing to make v1beta1 and realize a day too late that it was completely broken because of some stupid mistake we made, don't you think? Especially considering how good the v1alpha1 is.

Reply:
> I don't know if we are going to have a v1beta2. It's up in the air, and this work does not change that.

I mean that if we promote the current v1alpha1 to v1beta1 without changes, then we take the path to also do v1beta2, since we want to implement the changes discussed in this proposal.

> The reality is: we won't know the downsides of any API until people start using it. If we release v1alpha2 now, we'll need to wait months for everyone to switch. While advertising our API as not suitable for production.

Fair enough. I'm open to releasing the current v1alpha1 as v1beta1 (with the caveat about v1beta2 coming up). What I find embarrassing is taking our current production grade API that is labeled as alpha, make changes to it and release as beta, then find out it was broken due to the changes we did. It would be less embarrassing to let it take time and make sure it is tested under the alpha label. But like mentioned before, I'm fine with promoting v1alpha1 as beta directly with the goal of signaling to consumers that this is production ready.

> Then release v1beta1 and wait months more. Then someone comes and says "oh wait, I have a brilliant idea!" :)

The whole idea with this is that two API versions can be identical. If v1alpha2 turns out good, we just release it as v1beta1 when we feel it has been properly tested. No need to then wait more months after that release since it is the exact same as the previous.

To summarize my thoughts

- :heavy_check_mark: Promote v1alpha1 and release it unchanged as v1beta1. Then start working on v1beta2.
- :heavy_check_mark: Create a new v1alpha2 for implementing changes. Can be released as v1alpha2 or v1beta1 depending on how we feel about it.
- :heavy_minus_sign: Create a new v1beta1 for implementing changes. IMO unnecessarily bold move but I can accept it.
- :x: Make changes to v1alpha1. Unacceptable. This is after all the current API used in production. It must stay as it is.

Reply:
> I mean that if we promote the current v1alpha1 to v1beta1 without changes

I don't think this has been proposed though? Or is it what you would like to do? Making two versions in a short period of time may be confusing.

> find out it was broken due to the changes we did

Could you elaborate what you mean by broken and why it suddenly has become broken (or anyhow more broken than it is)? Most changes here are renaming of fields. The only functional change - the deprecation of MD5 - is essentially forced on us by upstream.

> release it as v1beta1 when we feel it has been properly tested

When is this moment going to happen? What's the definition of "properly tested" going to be? Will anyone even use the new version if they know that v1beta1 is coming soon?

We need to be very careful with assuming that anything we produce will be carefully tested in the near future. We've made this mistake in the Ironic community in the past. We kept producing release candidates that nobody really tested become everyone was waiting for a production-ready release. I'm afraid if we say "v1alpha2 is a candidate for the future v1beta1", the consumers will simply wait for v1beta1 (and then complain if it does not work). Same situation with the proposal of promoting v1alpha1 to v1beta1: why would anyone use that when a v1beta2 is coming soon?

Reply:
> I don't think this has been proposed though? Or is it what you would like to do?

It is my suggestion for how to solve the "Our software is production ready and has been for long time. We just never advertised it, and I want to fix it as soon as possible, not in several months." issue without compromising quality by rushing the new version.

> Could you elaborate what you mean by broken and why it suddenly has become broken (or anyhow more broken than it is)? Most changes here are renaming of fields. The only functional change - the deprecation of MD5 - is essentially forced on us by upstream.

It would be the first time we do this in BMO, right? I always expect issues then and would prefer to proceed with caution.

> When is this moment going to happen? What's the definition of "properly tested" going to be? Will anyone even use the new version if they know that v1beta1 is coming soon?

Maybe a month? We would definitely use it internally and I hope you too. The main goal for me would be to see that it actually works outside of the CI system and that the release process works so that the artifacts are usable after the release.

> Same situation with the proposal of promoting v1alpha1 to v1beta1: why would anyone use that when a v1beta2 is coming soon?

It is two different things:

1. Show that the project is ready for production (I didn't think this was urgent but you seem to think so)
2. Work on a new release including changes to the API (which should not be rushed in my opinion)

If 1 is not important we can skip it and then there is only the new API to care about. I still suggest labeling it as v1alpha2 in that case because it is the first time we do this with BMO. But if the community wants to do v1beta1 immediately I will accept it. It just seems unnecessarily "bold" with basically nothing to gain. On the other hand if we do v1alpha2 it would require basically no effort to promote it to v1beta1 later and then we avoid (most of) the risk of screwing it up.

Reply:
> without compromising quality by rushing the new version
> It just seems unnecessarily "bold"
> we avoid (most of) the risk of screwing it up

You keep implying some sort of quality regression stemming from this work without providing any reasoning. We have modified API in the past already, even though in a compatible way. Any change to BMO has a risk of breaking consumers.

If we create v1beta1 now, it probably will work (although the webhook part is the easiest to do wrongly). But how does it prevent us from "compromising quality" when doing v1beta2?

> Maybe a month? We would definitely use it internally and I hope you too. 

A month is not enough for us to switch - we have too many things going on. Knowing another version is coming, I may not even bother with switching because it's not free, and my time is limited.

> On the other hand if we do v1alpha2 it would require basically no effort to promote it to v1beta1 later 

It's not "no effort". We need to update the code, we need to update the consumers, we need to set a sunset date for v1alpha2.

In the end, I can agree to v1alpha1 -> v1alpha2 -> v1beta1 just for the sake of ending this argument, but I still see this as largely pointless. I would like to hear from other community members though, so that I don't have to change this plan again.

Reply:
Alright let's not drag this out more than necessary. Thank you for the patience :wink: We have discussed it in the team and going for v1beta1 is fine. I'm not understanding fully the need for this but I don't have a strong opinion so to end this thread, go with v1beta1. :slightly_smiling_face: 

There were some additional concerns/comments/ideas though, so expect a few more comments soonish from @kashifest and @Rozzii !